core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -417,7 +417,7 @@ private[log] class Cleaner(val id: Int,
     txnIndexFile.delete()
 
     val startOffset = segments.head.baseOffset
-    val records = FileRecords.open(logFile, false, log.initFileSize(), log.config.preallocate)
+    val records = FileRecords.open(logFile, false, log.initFileSize(), false)
```

Comment:
@huxihx : Wow. Great finding. Thanks. It does seem that we missed trimming the log in the log cleaner when preallocation is enabled. We probably want to keep FileRecords in the preallocate mode if it's enabled since it's possible for most records to be preserved after cleaning. What we can do is to follow what we do in Log.roll() by calling records.trim() before cleaned.flush(). It would also be useful to add a simple unit test  for this case (e.g., trigger a cleaning with no filtering and verify the cleaned segment size is trimmed).